// @mui/x-data-grid-premium/node/hooks/features/aggregation/createAggregationLookup.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createAggregationLookup = void 0;

var _xDataGridPro = require("@mui/x-data-grid-pro");

var _gridAggregationUtils = require("./gridAggregationUtils");

var _gridAggregationSelectors = require("./gridAggregationSelectors");

const getAggregationCellValue = ({
  apiRef,
  groupId,
  field,
  aggregationFunction,
  aggregationRowsScope
}) => {
  const rowTree = (0, _xDataGridPro.gridRowTreeSelector)(apiRef);
  const filteredRowsLookup = (0, _xDataGridPro.gridFilteredRowsLookupSelector)(apiRef);
  let rowIds; // TODO: Add custom root id

  if (groupId === '') {
    rowIds = (0, _xDataGridPro.gridRowIdsSelector)(apiRef).filter(rowId => !rowTree[rowId].isAutoGenerated);
  } else {
    rowIds = apiRef.current.getRowGroupChildren({
      groupId
    });
  }

  const values = [];
  rowIds.forEach(rowId => {
    if (aggregationRowsScope === 'filtered' && filteredRowsLookup[rowId] === false) {
      return;
    }

    values.push(apiRef.current.getCellValue(rowId, field));
  });
  return aggregationFunction.apply({
    values
  });
};

const getGroupAggregatedValue = ({
  groupId,
  apiRef,
  aggregationRowsScope,
  aggregatedFields,
  aggregationRules,
  position
}) => {
  const groupAggregationLookup = {};

  for (let j = 0; j < aggregatedFields.length; j += 1) {
    const aggregatedField = aggregatedFields[j];
    const columnAggregationRules = aggregationRules[aggregatedField];
    groupAggregationLookup[aggregatedField] = {
      position,
      value: getAggregationCellValue({
        apiRef,
        groupId,
        field: aggregatedField,
        aggregationFunction: columnAggregationRules.aggregationFunction,
        aggregationRowsScope
      })
    };
  }

  return groupAggregationLookup;
};

const createAggregationLookup = ({
  apiRef,
  aggregationFunctions,
  aggregationRowsScope,
  getAggregationPosition
}) => {
  const aggregationRules = (0, _gridAggregationUtils.getAggregationRules)({
    columnsLookup: (0, _xDataGridPro.gridColumnLookupSelector)(apiRef),
    aggregationModel: (0, _gridAggregationSelectors.gridAggregationModelSelector)(apiRef),
    aggregationFunctions
  });
  const aggregatedFields = Object.keys(aggregationRules);

  if (aggregatedFields.length === 0) {
    return {};
  }

  const aggregationLookup = {};
  const rowIds = (0, _xDataGridPro.gridRowIdsSelector)(apiRef);
  const rowTree = (0, _xDataGridPro.gridRowTreeSelector)(apiRef);

  for (let i = 0; i < rowIds.length; i += 1) {
    var _node$children;

    const rowId = rowIds[i];
    const node = rowTree[rowId];
    const hasChildren = (_node$children = node.children) == null ? void 0 : _node$children.some(childId => {
      var _rowTree$childId$posi;

      return ((_rowTree$childId$posi = rowTree[childId].position) != null ? _rowTree$childId$posi : 'body') === 'body';
    });

    if (hasChildren) {
      const position = getAggregationPosition(node);

      if (position != null) {
        aggregationLookup[rowId] = getGroupAggregatedValue({
          groupId: rowId,
          apiRef,
          aggregatedFields,
          aggregationRowsScope,
          aggregationRules,
          position
        });
      }
    }
  } // TODO: Add custom root id


  const position = getAggregationPosition(null);

  if (position != null) {
    aggregationLookup[''] = getGroupAggregatedValue({
      groupId: '',
      apiRef,
      aggregatedFields,
      aggregationRowsScope,
      aggregationRules,
      position
    });
  }

  return aggregationLookup;
};

exports.createAggregationLookup = createAggregationLookup;
